### PR TITLE
feat(Notification): no resize for email attachment

### PR DESCRIPTION
**Pull request by stonebuzz**
When image is added to email as attachment (see ```Add documents into ticket notifications``` option from notification configuration) 

Image is resized as mail context (```400px``` / ```300px```)

In this context ```mailattachment```  the image must not be resized

This Pr fix this

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !29869

### CODE REVIEW

File: src/NotificationEventMailing.php [hunk 1]
```diff
@@ -496,7 +496,7 @@ private static function attachDocuments(GLPIMailer $mmail, array $documents_ids)
             if (Document::isImage($path)) {
                 $path = Document::getImage(
                     $path,
-                    'mail'
+                    'mailattachment'
```
Comment (cedric-anne) [inline src/NotificationEventMailing.php:499]: Is the `mail` context still used ?

**Review by cedric-anne — COMMENTED** [review]
This limitation is here from a long time ago, and removing it may beak email sending, due to high memory consumption of the phpmailer way to handle attachements (see #13511).

I agree to remove this limitation, but not before GLPI 10.1. Also, we should do some benchmarks to ensure that big attachements will not consume too much memory.

File: src/NotificationEventMailing.php (hunk 1 above)
Comment (stonebuzz) [reply src/NotificationEventMailing.php:499]: Yes in ```NotificationEventMailing``` (2 times)

Comment (cedric-anne) [reply src/NotificationEventMailing.php:499]: If we decide to not compress images anymore in notifications, we should do it for all images. In this case, the `Document::getImage()` method will not be used anymore and could be deprecated.

Comment (stonebuzz) [reply src/NotificationEventMailing.php:499]: My only suggestion is that images in attachments should not be compressed.

I'm fine with resizing the email to make it easier to read

**Comment by orthagh** [comment]
After irl discussion:
 - ok for the change
 - target main branch
 - bench the changes

File: src/Document.php [hunk 2]
```diff
@@ -1805,38 +1805,21 @@ public static function isImage($file)
...
      *
      * @param string  $path    Original path
-     * @param string  $context Context
      * @param integer $mwidth  Maximal width
      * @param integer $mheight Maximal height
      *
      * @return string Image path on disk
      */
-    public static function getImage($path, $context, $mwidth = null, $mheight = null)
+    public static function getImage($path, $mwidth = null, $mheight = null)
```
Comment (cedric-anne) [inline src/Document.php:1819]: I would be safer to keep the previous `getImage()` method signature, deprecate it, and declare a new `public static function getResizedImagePath(string $path, int $width, int $height): string` method.

File: src/NotificationEventMailing.php [hunk 3]
```diff
@@ -283,10 +283,7 @@ public static function send(array $data)
                                 trigger_error('Failed to add document ' . $doc->fields['filepath'] . ' to mail: file not found', E_USER_WARNING);
                                 continue;
                             }
-                            $image_path = Document::getImage(
-                                GLPI_DOC_DIR . "/" . $doc->fields['filepath'],
-                                'mail'
-                            );
+                            $image_path = Document::getImage(GLPI_DOC_DIR . "/" . $doc->fields['filepath']);
```
Comment (cedric-anne) [inline src/NotificationEventMailing.php:286]: If there is no need to resize, then there is no need to call any method.
```suggestion
                            $image_path = GLPI_DOC_DIR . "/" . $doc->fields['filepath'];
```

File: src/NotificationEventMailing.php [hunk 4]
```diff
@@ -518,10 +514,7 @@ private static function attachDocuments(Email $mail, array $documents_ids)
             }
             $path = GLPI_DOC_DIR . "/" . $document->fields['filepath'];
             if (Document::isImage($path)) {
-                $path = Document::getImage(
-                    $path,
-                    'mail'
-                );
+                $path = Document::getImage($path);
             }
```
Comment (cedric-anne) [inline src/NotificationEventMailing.php:518]: This whole piece of code can be removed.

**Review by cedric-anne — CHANGES_REQUESTED** [review]
`$context` parameter could be removed from `Document::send()` method and from its usage in `document.send.php` script.

File: src/Document.php (hunk 2 above)
Comment (stonebuzz) [reply src/Document.php]: done !

File: src/NotificationEventMailing.php (hunk 3 above)
Comment (stonebuzz) [reply src/NotificationEventMailing.php]: done !

File: src/NotificationEventMailing.php (hunk 4 above)
Comment (stonebuzz) [reply src/NotificationEventMailing.php]: done !

**Comment by stonebuzz** [comment]
before adapt test
max ```memory_get_usage```        => 42.1 Mio
max ```memory_get_usage(true)```  => 46 Mio
max ```memory_get_peak_usage()``` => 46 Mio

after adadpt test
max ```memory_get_usage```        => 42.1 Mio
max ```memory_get_usage(true)```  => 46 Mio
max ```memory_get_peak_usage()``` => 51 Mio

**Comment by stonebuzz** [comment]
Change bench method

Before (without large  file / img)

2.78 Mio
2.96 Mio
2.97 Mio
2.97 Mio
2.97 Mio
2.97 Mio
2.97 Mio
2.97 Mio
2.98 Mio
2.98 Mio
2.98 Mio
2.99 Mio
2.99 Mio
2.99 Mio
3 Mio
2.91 Mio
2.83 Mio
2.76 Mio
2.67 Mio
2.6 Mio
2.52 Mio
2.43 Mio
2.36 Mio
2.28 Mio
2.19 Mio
2.12 Mio
1.92 Mio
1.83 Mio
1.75 Mio
1.68 Mio

Before (with large  file / img)

11.28 Mio
10.73 Mio
10.62 Mio
10.5 Mio
8.59 Mio
8.44 Mio
8.31 Mio
8.19 Mio
8.06 Mio
7.92 Mio
7.79 Mio
7.66 Mio
7.45 Mio
7.31 Mio
7.17 Mio
7.08 Mio
7.01 Mio
6.93 Mio
6.85 Mio
6.77 Mio
6.7 Mio
6.61 Mio
6.53 Mio
6.46 Mio
6.37 Mio
6.29 Mio
6.09 Mio
6.01 Mio
5.93 Mio
5.85 Mio

**Review by cedric-anne — APPROVED** [review]
I reverted the changes made on test because they were not testing the real consumption usage. Indeed, we use a transport mockup that does not actually sends anything.

Here is the results of tests I made manually.
- Sending one notification with a 1Mb random TXT attachement takes about 3.5Mb of memory.
- Sending one notification with 10 x 1Mb random TXT attachements takes about 15Mb of memory.
- Sending 30 notifications with 10 x 1Mb random TXT attachements takes about 16Mb of memory.
- Sending one notification with 10 x 15Mb random TXT attachements, takes about 215Mb of memory.

In summary, memory usage is really optimized on symfony/mailer component due to intensive usage of streams. Once a notification is sent, memory is freed, so, unless the administrator puts a really low PHP memory limit, there is no problem to send big file attachements.

**Review by orthagh — APPROVED** [review]

**Review by trasher — APPROVED** [review]
